doc/examples/example5.md

```diff
@@ -0,0 +1,397 @@
+# Asymmetric Crypto API {#example5}
+
+# Common Interface
+
+## Encryption Interface
+
+All contexts that support en- / decryption are implementing an `encrypt()` or `decrypt()` method resprectively. The classes `mococrw::EncryptionCtx` and `mococrw::DecryptionCtx` provide a pure virtual interface to these methods:
+<!-- TODO for when refactoring Examples: Generate such code snippets from code-->
+\code{.cpp}
+class EncryptionCtx {
+    ...
+    virtual std::vector<uint8_t> encrypt(const std::vector<uint8_t>& message) = 0;
+};
+
+class DecryptionCtx {
+    ...
+    virtual std::vector<uint8_t> decrypt(const std::vector<uint8_t>& message) = 0;
+}
+\endcode
+That is, these classes can be used to access a en-/decryption context in a generic manner.
+
+## Signature Interface
+
+All contexts that support signing implement a method to sign pre-hashed message digests (`signDigest()`) or unhashed messages (`signMessage()`) or both. Please note that if a message digest shall be signed (or verified) it is expected that the client provides the digest of the message to be signed (i.e. performs the hashing on his own). On the contrary, signing a message means that the message digest will be calculated and signed then. The classes `mococrw::DigestSignatureCtx`, `mococrw::MessageSignatureCtx`, `mococrw::DigestVerificationCtx` and `mococrw::MessageVerificationCtx` provide a pure virtual interface to these methods:
+
+\code{.cpp}
+class DigestSignatureCtx {
+    ...
+    virtual std::vector<uint8_t> signDigest(const std::vector<uint8_t> &messageDigest) = 0;
+};
+
+class MessageSignatureCtx {
+    ...
+    virtual std::vector<uint8_t> signMessage(const std::vector<uint8_t> &message) = 0;
+};
+
+class DigestVerificationCtx {
+    ...
+    virtual void verifyDigest(const std::vector<uint8_t> &signature,
+                              const std::vector<uint8_t> &digest) = 0;
+};
+
+class MessageVerificationCtx {
+    ...
+    virtual void verifyMessage(const std::vector<uint8_t> &signature,
+                               const std::vector<uint8_t> &message) = 0;
+};
+\endcode
+
+#### Example
+If a context supports both, signing messages and message digest, the following code snippets are equal. For simplicity it is assumes that SHA256 is used to create the signed hash.
```

Comment:
"it is assumes" is not good Engrish. I propose we turn this into "it is assumed" or "we assume".